test/extended/etcd/helpers/helpers.go

```diff
@@ -247,7 +246,7 @@ func IsCPMSActive(ctx context.Context, t TestingT, cpmsClient machinev1client.Co
 // EnsureReadyReplicasOnCPMS checks if status.readyReplicas on the cluster CPMS is n
 // this effectively counts the number of control-plane machines with the provider state as running
 func EnsureReadyReplicasOnCPMS(ctx context.Context, t TestingT, expectedReplicaCount int, cpmsClient machinev1client.ControlPlaneMachineSetInterface) error {
-	waitPollInterval := 5 * time.Second
+	waitPollInterval := 2 * time.Second
 	waitPollTimeout := 18 * time.Minute
 	t.Logf("Waiting up to %s for the CPMS to have status.readyReplicas = %v", waitPollTimeout.String(), expectedReplicaCount)
```

Comment:
This comment is for L259-266 (since Github [won't let me comment on unchanged lines](https://github.com/github/roadmap/issues/347))
```Go
if cpms.Status.ReadyReplicas != int32(expectedReplicaCount) {
			t.Logf("expected %d ready replicas on CPMS, got: %v,", expectedReplicaCount, cpms.Status.ReadyReplicas)
	return false, nil
}

t.Logf("CPMS has reached the desired number of ready replicas: %v,", cpms.Status.ReadyReplicas)
return true, nil
```

Previous discussion `EnsureReadyReplicasOnCPMS ` https://github.com/openshift/origin/pull/27702#discussion_r1106171808

> deads2k: do you care about this condition or do you care this AND that you have exactly the right number of total nodes AND that the desired replicas is also expectedReplicaCount?

> hasbro17: We care about this condition because we want the test to ensure that a new machine has successfully been created and is ready, which means the associated node has also been created and is running.
IIUC ReadyReplicas should cover both conditions:
> https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/main/pkg/controllers/controlplanemachineset/status.go#L95-L96
> https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/main/pkg/machineproviders/types.go#L38-L40
> I think the desired status.replicas will only tell us if the machine was created, and not whether it became running with a node attached.

> deads2k: if expected does not match the desired (spec), the return further down looks incorrect.
> also, this looks like the correct spot to double check that the number of nodes is actually correct.

> deads2k (slack): EnsureReadyReplicasOnCPMS doesn't appear to be sure
> 1. that .spec.replicas matches .status.readyreplicas
> 2. that the number of nodes in the API is actually correct

Addressing the above two points:
### 2. that the number of nodes in the API is actually correct

As I pointed out above, checking `cpms.Status.ReadyReplicas` will also indirectly check that the number of nodes are correct. The CPMSO updates `status.ReadyReplicas` with machines that are only [counted as ready](https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/main/pkg/controllers/controlplanemachineset/status.go#L95-L96) when [its node is also Ready](https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/main/pkg/machineproviders/types.go#L38-L40)

We actually caught a bug in this behavior with our test that has since been verified and fixed.
https://issues.redhat.com/browse/OCPBUGS-7989
https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/171

### 1. that `.spec.replicas` matches `.status.readyReplicas`

The purpose of `EnsureReadyReplicasOnCPMS` in this test is not to check if  `spec.replicas` matches `status.readyReplicas`.
It is to check whether `status.readyReplicas` is whatever we expect it to be at a given point in the test i.e `expectedReplicaCount`. 

While `spec.replicas` [will always be 3](https://github.com/openshift/cluster-control-plane-machine-set-operator/blob/main/docs/user/installation.md#anatomy-of-a-controlplanemachineset) it is expected behavior from the CPMSO that `status.readyReplicas` will surge up to `4` as we scale-up and that's what we check for in step 2 of the test with:
```Go
 err = scalingtestinglibrary.EnsureReadyReplicasOnCPMS(ctx, g.GinkgoT(), 4, cpmsClient)
```

And then in step 3 after scale-down we expect it to come back down to `3`:
```Go
 err = scalingtestinglibrary.EnsureReadyReplicasOnCPMS(ctx, g.GinkgoT(), 3, cpmsClient)
```

@deads2k Does that address your previous comments?

Reply:
>We actually caught a bug in this behavior with our test that has since been verified and fixed.
https://issues.redhat.com/browse/OCPBUGS-7989
https://github.com/openshift/cluster-control-plane-machine-set-operator/pull/171

This sounds like a really strong argument for adding the verification I've listed.

Reply:
>The purpose of EnsureReadyReplicasOnCPMS in this test is not to check if spec.replicas matches status.readyReplicas.
It is to check whether status.readyReplicas is whatever we expect it to be at a given point in the test i.e expectedReplicaCount.

Accepted.  There is some test to be sure we eventually converge?

Reply:
> This sounds like a really strong argument for adding the verification I've listed.

Fair point. The CPMSO should have test cases for this with the recent fix, but we can add a check for the node count here to ensure we don't regress.

> There is some test to be sure we eventually converge?

Not quite. Indirectly we do check that it goes down to 3. And `spec.replicas` is supposed to be immutable but that could change in the future.
So based on that, we'll add a check at the end to ensure they both are the same after scale-down if that guarantee ever changes.

Reply:
Alright, with the latest updates the test should now be checking that the number of nodes are correct, as well as ensuring that `spec.replicas` and `status.readyReplicas` converge at the end of the test.